src/msg/consumer/types.go

```diff
@@ -132,17 +132,58 @@ type MessageProcessor interface {
 	Close()
 }
 
-// NewMessageProcessorFn creates a new MessageProcessor scoped to a single connection. Messages are processed serially
-// in a connection.
-type NewMessageProcessorFn func() MessageProcessor
+// MessageProcessorPool returns MessageProcessors.
+type MessageProcessorPool interface {
+	// Get returns a MessageProcessor.
+	Get() MessageProcessor
+	// Put returns the MessageProcessor.
+	Put(mp MessageProcessor)
+	// Close the pool.
+	Close()
+}
+
+// SingletonMessageProcessor returns a MessageProcessorPool that shares the same MessageProcessor for all users. The
+// MessageProcessor is closed when the pool is closed.
+func SingletonMessageProcessor(mp MessageProcessor) MessageProcessorPool {
```

Comment:
I don't see this referenced anywhere? Am I missing something?

Reply:
it already existed https://github.com/m3db/m3/blob/master/src/cmd/services/m3coordinator/server/m3msg/config.go#L89

The function signature just changed

Reply:
Ah I see - thanks